### PR TITLE
Add a distributed lock

### PR DESCRIPTION
**Pull request by erikjohnston**
This adds a simple best effort locking mechanism that works cross workers.

### CODE REVIEW

File: synapse/storage/databases/main/lock.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                    keyvalues={
+                        "lock_name": lock_name,
+                        "lock_key": lock_key,
+                    },
+                    values={},
+                    insertion_values={
+                        "token": token,
+                        "last_renewed_ts": self._clock.time_msec(),
+                    },
+                )
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:165]: this is missing a value for `instance_name` I think (and hence CI is failing)

File: changelog.d/10269.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Add a distributed lock.
```
Comment (richvdh) [inline changelog.d/10269.misc:1]:
```suggestion
Add a distributed lock implementation.
```

File: synapse/storage/databases/main/lock.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        self._reactor = hs.get_reactor()
+        self._instance_name = hs.get_instance_id()
+
+        # A map from `(lock_name, lock_key)` to the token of any locks that we
+        # think we currently hold.
+        self._live_tokens: Dict[Tuple[str, str], str] = {}
+
+        # When we shut down we want to remove the locks. Technically this can
+        # lead to a race, as we may drop the lock while we are still processing.
+        # However, a) it should be a small windown, b) the lock is best effort
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:68]:
```suggestion
        # However, a) it should be a small window, b) the lock is best effort
```

File: synapse/storage/databases/main/lock.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                sql = """
+                    INSERT INTO worker_locks (lock_name, lock_key, instance_name, token, last_renewed_ts)
+                    VALUES (?, ?, ?, ?, ?)
+                    ON CONFLICT (lock_name, lock_key)
+                    DO UPDATE SET
+                        token = EXCLUDED.token,
+                        instance_name = EXCLUDED.instance_name,
+                        last_renewed_ts = EXCLUDED.last_renewed_ts
+                    WHERE
+                        worker_locks.last_renewed_ts < ?
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:109]: I had to stare at this for a while; I failed to grok the WHERE applied only to the UPDATE.

```suggestion
                    DO UPDATE 
                        SET
                            token = EXCLUDED.token,
                            instance_name = EXCLUDED.instance_name,
                            last_renewed_ts = EXCLUDED.last_renewed_ts
                        WHERE
                            worker_locks.last_renewed_ts < ?
```

File: synapse/storage/databases/main/lock.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+            desc="is_lock_still_valid",
+        )
+        # If there is an entry and it has not timed out, then we still hold the lock
+        if (
+            last_renewed_ts is not None
+            and self._clock.time_msec() - _LOCK_TIMEOUT_MS < last_renewed_ts
+        ):
+            return True
+
+        return False
```
Comment (clokep) [inline synapse/storage/databases/main/lock.py:210]:
```suggestion
        return (
            last_renewed_ts is not None
            and self._clock.time_msec() - _LOCK_TIMEOUT_MS < last_renewed_ts
        )
```

File: synapse/storage/databases/main/lock.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            keyvalues={
+                "lock_name": lock_name,
+                "lock_key": lock_key,
+                "token": token,
+            },
+            updatevalues={"last_renewed_ts": self._clock.time_msec()},
+            desc="renew_lock",
+        )
+
+        self._live_tokens.pop((lock_name, lock_key), None)
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:221]: this looks wrong? should be in `_drop_lock`, no?

File: tests/storage/databases/main/test_lock.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+        # We simulate the process getting stuck by cancelling the looping call
+        # that keeps the lock active.
+        lock._looping_call.stop()
+
+        # Wait for the lock to timeout.
+        self.reactor.advance(2 * _LOCK_TIMEOUT_MS / 1000)
+
+        lock2 = self.get_success(self.store.try_acquire_lock("name", "key"))
+        self.assertIsNotNone(lock2)
```
Comment (clokep) [inline tests/storage/databases/main/test_lock.py:84]: We should assert that the callback from `lock` now returns `False`, I think.

File: synapse/storage/databases/main/lock.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    the lock is still valid, which can be used to bail out of processing if e.g.
+    processing work in a loop.
+
+    For example:
+
+        lock = await self.store.try_acquire_lock(...)
+        if not lock:
+            return
+
+        async with lock() as still_valid:
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:250]: is this correct? should it be:

```suggestion
        async with lock as still_valid:
```

Comment (richvdh) [reply synapse/storage/databases/main/lock.py:250]: also, this business of returning a callback from the context manager feels pretty confusing. Maybe just call `lock.is_lock_still_valid` from inside the loop.

File: synapse/storage/databases/main/lock.py [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        await self._store._renew_lock(self._lock_name, self._lock_key, self._token)
+
+    async def is_lock_still_valid(self) -> bool:
+        """Check if the lock is still held by us"""
+        return await self._store._is_lock_still_valid(
+            self._lock_name, self._lock_key, self._token
+        )
+
+    async def __aenter__(self) -> Callable[[], Awaitable[bool]]:
+        return self.is_lock_still_valid
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:290]: maybe assert that we have not dropped the lock? (ie, that we are not reusing the context manager)

File: synapse/storage/databases/main/lock.py [hunk 10]
```diff
@@ -200,14 +200,10 @@ async def _is_lock_still_valid(
             allow_none=True,
             desc="is_lock_still_valid",
         )
-        # If there is an entry and it has not timed out, then we still hold the lock
```
Comment (clokep) [inline synapse/storage/databases/main/lock.py]: Oops I don't think I meant to delete this comment in my suggestion.

File: synapse/storage/databases/main/lock.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+
+        return False
+
+    def __del__(self) -> None:
+        if not self._dropped:
+            # We should not be dropped without the lock being released (unless
+            # we're shutting down), but if we are then let's at least stop
+            # renewing the lock.
+            if self._looping_call.running:
+                self._looping_call.stop()
```
Comment (richvdh) [inline synapse/storage/databases/main/lock.py:312]: given the timer holds a reference to `renew`, and `renew` holds a reference to `self`... I suspect this is never going to be reached. You could maybe make the renew task a function with minimal closure.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: synapse/storage/databases/main/lock.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/lock.py:221]: oh lord, yes

File: synapse/storage/databases/main/lock.py (hunk 9 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/lock.py:305]: good point

**Review by richvdh — APPROVED** [review]
lgtm

File: synapse/storage/databases/main/lock.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+
+        lock = await self.store.try_acquire_lock(...)
+        if not lock:
+            return
+
+        async with lock:
+            for item in work:
+                await process(item)
+
+                if not await is_still_valid():
```
Comment (clokep) [inline synapse/storage/databases/main/lock.py:254]: I think?

```suggestion
                if not await lock.is_still_valid():
```

Comment (richvdh) [reply synapse/storage/databases/main/lock.py]: yup, thanks

**Pull request merged**